clap_generate/src/man.rs

```diff
@@ -0,0 +1,287 @@
+        // by a newline the alternative is to add \& at the end of the line
+        buffer.push_str(".ss \\n[.ss] 0\n");
+        // Disable hyphenation
+        buffer.push_str(".nh\n");
+        // Disable justification (adjust text to the left margin only)
+        buffer.push_str(".ad l\n");
+
+        self.render_about(&mut buffer);
+        self.render_description(&mut buffer);
+        self.render_synopsis(&mut buffer);
+
+        self.render_options(&mut buffer);
+        self.render_subcommands(&mut buffer);
+
+        self.render_custom_sections(&mut buffer);
+
+        self.render_version(&mut buffer);
+        self.render_authors(&mut buffer);
+
+        w!(buf, buffer.as_bytes());
+    }
+
+    fn render_header(&self, buf: &mut String) {
+        write(
+            buf,
+            format_args!(
+                ".TH {} {}",
+                escape(self.app.get_name()).to_uppercase(),
+                self.section
+            ),
+        )
+        .unwrap();
+
+        // TODO: how to handle date.
+        // match date {
+        //     Some(date) => write(out, format_args!(r#" "{}""#, date)).unwrap(),
+        //     None => out.push_str(r#" """#),
+        // }
+
+        buf.push_str(r#" """#);
+
+        match self.app.get_version() {
+            Some(version) => write(
+                buf,
+                format_args!(r#" "{} {}""#, self.app.get_name(), version),
+            )
+            .unwrap(),
+            None => buf.push_str(r#" """#),
+        }
+
+        match self.manual.as_ref() {
+            Some(manual) => write(buf, format_args!(r#" "{}""#, manual)).unwrap(),
+            None => buf.push_str(r#" """#),
+        }
+
+        buf.push('\n');
+    }
+
+    fn render_about(&self, buf: &mut String) {
+        buf.push_str(".SH \"NAME\"\n");
+        buf.push_str(&bold(self.app.get_name()));
+
+        if let Some(about) = self.app.get_about() {
+            buf.push_str(r" \- ");
+            buf.push_str(&escape(about));
+        }
+
+        buf.push('\n');
+    }
+
+    fn render_description(&self, buf: &mut String) {
+        if let Some(about) = self.app.get_long_about() {
+            buf.push_str(".SH \"DESCRIPTION\"\n");
+            buf.push_str(about);
+            buf.push('\n');
+        }
+    }
+
+    fn render_synopsis(&self, buf: &mut String) {
+        buf.push_str(".SH \"SYNOPSIS\"\n");
+        buf.push_str(&italic(self.app.get_name()));
+        buf.push(' ');
+
+        for opt in self.app.get_arguments() {
+            match (opt.get_short(), opt.get_long()) {
+                (Some(short), Some(long)) => {
+                    write(buf, format_args!("[-{}|--{}] ", short, long)).unwrap()
+                }
+                (Some(short), None) => write(buf, format_args!("[-{}] ", short)).unwrap(),
+                (None, Some(long)) => write(buf, format_args!("[--{}] ", long)).unwrap(),
+                (None, None) => (),
+            }
+        }
+
```

Comment:
Could you point me to where I can find this? From searching through docs.rs I couldn't find a good function to use for this.

Reply:
`App::render_usage` though it won't let you do any inline formatting.  If thats a problem, we can try to decide which route to go.

Reply:
If we were to use something like this we'd need to have it be more modular, the `render_usage()` methods output is not really usable in a manpage as is since there is no formatting and it is missing all the options.

Reply:
`DontCollapseArgsInUsage` helps some but not all the way.

Is it recommended to not collapse anything for man pages?  Seems like that could get overly verbose

Looks like we are italicizing the name but nothing else.  On the other hand, there is a lot of relevant logic in our usage generator, like more thorough detection of required arguments, etc.  I'm mixed on which way to go.

Reply:
I'm not sure what the preference is, `git` does some line splitting while `fd` squashes all short options together (`fd [-HIEsiaLp0hV] [-d depth]` and only shows the short name for options.

Reply:
I guess we can stick to git's precedence for now and see how the feedback turns out